### DISCUSSION TITLE
Why does sbt need native code?

### DISCUSSION THREAD

**Discussion by hrj** [sbt/sbt#6202] [Q&A] [answered]
Hi!

I was a little surprised to see in [this issue](https://github.com/sbt/sbt/issues/6162) that sbt uses native code.

I imagine that native code is necessary for trapping the `ctrl-c` signal to kill sbt tasks rather than sbt itself. The last time I checked, Java didn't provide a mechanism for trapping `ctrl-c`.

What else is native code used for in sbt? Perhaps we can find alternative jvm solutions to those workarounds.

thanks!

**Comment by eed3si9n** [comment, marked as answer]
We use native and native-adjacent code in sbt for various reasons, mainly we do so to compensate the lack of first-class feature from the JVM. Here are the areas that I can think of off the top of my head:

### Interprocess communication (IPC)

I created ipcsocket (https://github.com/sbt/ipcsocket) in 2018 that wraps simple IPC socket calls in macOS, Linux, and Windows. It's relatively common to use IPC socket (Unix domain socket on POSIX, and named pipe on Windows) for Node.JS implementing Language Server etc, and unfortunately JVM does not provide the feature. There's [JEP 380](https://openjdk.java.net/jeps/380) as proposal now, so maybe in the future we could rely on that. Why not use TCP socket? By design, TCP sockets will be available over network, which would open up sbt server to potential attacks. At least it would require encryption and authentication, basically like `ssh`.

### Filesystem event watch

Ethan Atkins created file tree views (https://github.com/swoval/swoval) in 2017 and was initially adopted in sbt as the best alternative to file watching implementation since `java.nio.file.WatchService` implementation is missing for macOS. File watch is used in sbt for triggered execution also known as tilde command `~`. Ethan joined the development of sbt from 2018 - 2020 and implemented various performance improvements using file tree views, including the recursive listing of files.

### Console user interface (CUI)

sbt offers rich console user interface capabilities such as history retention and tab completion by utilizing JLine, which also uses JAnsi. For REPL, Scala compiler implementations also use JLine 2 or 3 as well. Since REPLs run inside of sbt's JVM process, we need to be cognizant of how JLine is invoked with a REPL session to avoid conflicts.

### Thin client

sbt 1.4.0 introduced a thin client of sbt server called `sbtn`, which is implemented as Graal native image built for macOS, Linux, and Windows. The main motivation for providing a thin client in native image is to avoid JVM spin-up overhead (loading `*.class` and JIT). Having `sbtn` allows interaction in the order of 60 ~ 80ms. As this uses GraalVM as a mechanism to create native images, we will always have the equivalent JVM code for the thin client features.

### Capability matrix

> Perhaps we can find alternative jvm solutions to those workarounds.

Since the majority of the users and contributors use one of macOS, Linux, and Windows, I don't think it would be possible or desirable to replace the use of native code solutions with pure-Java alternatives. However, I think it would be a useful exercise to understand the capability matrix of `sbt`'s features and different operating systems. For example, if we manage to install `sbt` into a FreeBSD machine or a Raspberry Pi (ARM + Linux), are all features borked, or can we still get the basic feature to work?

**Reply by eatkins** [reply]
@eed3si9n covered the bases well. I'd add that the `~` command and jansi usages of native code are must have. Both are specific to either mac or windows which are common platforms and must be supported. It's unfortunate that platform architecture changes can break sbt but we can't provide core functionality without native code so all we can do is make timely releases with updated native code when necessary. It may be the case that future jdks support all of the functionality that sbt needs on all platforms or at least enough that we could drop native code usage in the future. Even if the jdk catches up, we would have to wait for an sbt 2 or 3 since 1.x is guaranteed to work with java 8.

The apple silicon issue in #6162 came up because as an enhancement, even when not in `~`, we watch build files for changes so that we can automatically reload the build if a project file changes. We thus tried to load the native code when sbt started up, causing a crash. This would not have happened on any non-osx platform because only osx uses custom native code for watch. We also generally have fallbacks if the native code cannot be loaded if the native code is just present as an enhancement (like the performance improvements of swoval for recursively listing the contents of directories).

Where possible sbt also uses jna rather than jni which prevents the need to distribute platform specific binaries. Unfortunately there is no jna version that works with apple silicon. Fortunately we had jni implementations of all the required native code for osx.

I guess future work would be to add a built in native code updater so that we could distribute native code in platform specific binary files that can be resolved like other libraries. We might have been able to avoid #6162 if we had something like that built into the launcher.

**Reply by eatkins** [reply]
I'd be curious about arm + linux as well. There is no arm64 implementation of swoval for linux or windows, but it should fall back to nio for those platforms. I'm pretty sure x86_64 freebsd works.
